rclcpp/include/rclcpp/node.hpp

```diff
@@ -305,16 +305,58 @@ class Node : public std::enable_shared_from_this<Node>
    *   name is invalid.
    * \throws rclcpp::exceptions::InvalidParameterValueException if initial
    *   value fails to be set.
+   * \throws rclcpp::exceptions::InvalidParameterTypeException
+   *   if the type of the default value or override is wrong.
    */
   RCLCPP_PUBLIC
   const rclcpp::ParameterValue &
   declare_parameter(
     const std::string & name,
-    const rclcpp::ParameterValue & default_value = rclcpp::ParameterValue(),
+    const rclcpp::ParameterValue & default_value,
     const rcl_interfaces::msg::ParameterDescriptor & parameter_descriptor =
     rcl_interfaces::msg::ParameterDescriptor(),
     bool ignore_override = false);
 
+  /// Declare and initialize a parameter, return the effective value.
+  /**
+   * Same as the previous one, but a default value is not provided and the user
+   * must provide a parameter override of the correct type.
+   *
+   * \param[in] name The name of the parameter.
+   * \param[in] type Desired type of the parameter, which will enforced at runtime.
+   * \param[in] parameter_descriptor An optional, custom description for
+   *   the parameter.
+   * \param[in] ignore_override When `true`, the parameter override is ignored.
+   *    Default to `false`.
+   * \return A const reference to the value of the parameter.
+   * \throws Same as the previous overload taking a default value.
+   * \throws rclcpp::exceptions::InvalidParameterTypeException
+   *   if an override is not provided or the provided override is of the wrong type.
```

Comment:
Wait so you're gonna throw `rclcpp::exceptions::InvalidParameterTypeException` when the user doesn't provide an override? Shouldn't we have a more precise error?

I guess the idea is that no override results in `NOT_SET` which will not match the ParameterType given? It's just surprising.

Reply:
> Shouldn't we have a more precise error?

Sounds reasonable to me.
I'm not sure about naming though, does `rclcpp::exceptions::NoOverrideProvided` sounds good?

> I guess the idea is that no override results in NOT_SET which will not match the ParameterType given? It's just surprising.

Yes, that was the idea.
`undeclare_parameter` will also throw a `InvalidParameterTypeException` exception, but with a better error message than this case.
I'm not sure if we also need a different exception for that one too https://github.com/ros2/rclcpp/blob/0e9be57f3cb9941fc96a606f7ada9a0481ee0627/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L560-L563.

Reply:
Hmm, I suppose the error type doesn't matter that much as long as the error message is clear, except if there's a case where they want to catch one but not the other error case.

In general, I'm in favor of more, and more specific, exception types.

> does rclcpp::exceptions::NoOverrideProvided sounds good?

`NoParameterOverrideProvided`? I don't feel strongly about the name. Just that the use of the existing exception type was a bit weird. Even if it was kind of correct.

Reply:
Fixed in https://github.com/ros2/rclcpp/pull/1522/commits/89a734e1d664ec0b2c5043fc997f126d43256acf